### PR TITLE
Add tests for scale_workers in aws_utils

### PR DESCRIPTION
**Pull request by matsujika**
[**The corresponding GCI task**](https://codein.withgoogle.com/dashboard/task-instances/5075263562448896/)

**Changes**:
- Add tests for `scale_workers` in `aws_utils.py`.

I referred to https://github.com/Cloud-CV/EvalAI/pull/2600 by @KhalidRmb

**CC**: @yashdusing @pushkalkatara @Ram81 @RishabhJain2018 @KhalidRmb

### CODE REVIEW

**Comment by nikochiko** [comment]
@takitsuse The Travis is passing but I don't think just changing the location of deploymentController param should be the solution. I was looking into it a little and found these docs for the version we are using: https://boto3.amazonaws.com/v1/documentation/api/1.9.88/reference/services/ecs.html#ECS.Client.create_service 
the `service_definition` json is passed to this method in our function. See: https://github.com/Cloud-CV/EvalAI/blob/master/apps/challenges/aws_utils.py#L347-L356
Can you please share a logic for its working if you have already? Or otherwise we could share our findings and discover what was causing the error.

P.S. Sorry for poking into this task, I was looking to claim another task for testing aws utils, so I hope you don't mind.

**Comment by nikochiko** [comment]
In fact, it is the task definition that is not supposed to have the `deploymentController` param, but it still works. See: https://boto3.amazonaws.com/v1/documentation/api/1.9.88/reference/services/ecs.html#ECS.Client.register_task_definition. I think we need to dig into this more. I will create another issue for this if you think appropriate?

**Comment by takitsuse** [comment]
I made that change in reference to https://github.com/Cloud-CV/EvalAI/pull/2596. So unfortunately I am ignorant about this issue but keen to solve it with you.

**CC**: @nsjcorps

**Comment by nikochiko** [comment]
Okay :+1:. I started a discussion about it with @nsjcorps on Gitter. We can
have a more formal discussion in an issue thread if even he doesn't know
about how the fix worked.

**Comment by takitsuse** [comment]
@nikochiko The test code is actually working as expected, so I assume `deployController` is ignored by `boto3`. I am going to bed so could you try deleting it?

**Comment by nsjcorps** [comment]
@nikochiko @takitsuse , As I told nikochiko, the issue was initially with the deployment controller being put in the wrong place. When that is resolved you should encounter more issues like the cluster name not being a cluster. I used mock to patch that. Which I can see that @takitsuse is using my code. I then mocked the ecs cluster to create the cluster.

**Comment by nsjcorps** [comment]
There were a bunch of issues I encountered even after the deploymentController issue was resolved

**Comment by nsjcorps** [comment]
I know how the fix works because....I made the fix, 😂

**Comment by takitsuse** [comment]
@nsjcorps Yeah but no issue I encountered after `deployController` issue is resolved.

**Comment by nsjcorps** [comment]
@takitsuse That's because you used my code and I have already resolved those issues

**Comment by nikochiko** [comment]
@takitsuse I thought the same about trying to remove the
"deploymentController" param from the task_definition. But as boto3 threw
an error about an extra argument for `create_service`, it should do the
same for `register_task_definition`. It should still be worth trying
though, so I will pull your branch and try some testing.

@nsjcorps yes it is working but I really don't understand how it worked. So
there are two issues I am facing:
1. Why boto3 raised an error when we added params according to [docs for
service_definition](
https://boto3.amazonaws.com/v1/documentation/api/1.9.88/reference/services/ecs.html#ECS.Client.create_service
).
2. Why boto3 didn't raise an error when we added the extra argument for
[task_definition](
https://boto3.amazonaws.com/v1/documentation/api/1.9.88/reference/services/ecs.html#ECS.Client.register_task_definition
).

It would be great if you linked a few docs to confirm your fix works
correctly. Otherwise, it may cause the whole aws_utils to work incorrectly,
like registering task definition when it is supposed to create a service.

@KhalidRmb it would be great if you can pitch in to solve the confusion.

P.S. the documentation I linked is for the version used in EvalAI i.e.
1.9.88

**Comment by takitsuse** [comment]
@nsjcorps I used only your changes that moves place of `deployController` and that cleared errors up. I mean, I saw you still had errors after you made that changes

**Comment by nsjcorps** [comment]
@takitsuse you also used my setup for the tests. That's where the errors are. The variable names are the same and so are the cluster names

**Comment by nsjcorps** [comment]
@nikochiko I didn't put the deployment Controller in the task definition. The task definition is completely different. If you trace the functions you will see what the task definition does

**Comment by nsjcorps** [comment]
I can see that you guys might be confused and it's a little hard to explain in here. I and sure of my changes though. Don't they work on your systems?

**Comment by takitsuse** [comment]
@nsjcorps I see.

**Comment by nsjcorps** [comment]
@takitsuse you understand now?

**Comment by takitsuse** [comment]
On local systems, the test on `aws_utils.py` works easily since https://github.com/Cloud-CV/EvalAI/pull/2524 and that unfortunately makes debugging way harder.

**Comment by nikochiko** [comment]
@nsjcorps @takitsuse Sorry for the confusion with the task_definition there. I thought you moved the deploymentController param into task_definition. So that makes it clearer. 
@nsjcorps Can you share a doc that says the deploymentController param should go into the VPC dict? I'm unable to find any.

**Comment by nsjcorps** [comment]
No. It works on all systems.

Test my code. Pull it on your system and see how it works

**Comment by nsjcorps** [comment]
The VPC dict is a dictionary that contains all the parameters that should not be in the "service" property. If you notice it uses **VPC dict which means it will automatically fill in all the values. The deployment controller is not supposed to be in the service definition. Hence, why it threw an error that it is not one of the valid types.

**Comment by takitsuse** [comment]
@nsjcorps I am still reading documentations, please wait.

**Comment by nikochiko** [comment]
Okay @nsjcorps I think I understand how your fix worked now. 
Refer: https://github.com/Cloud-CV/EvalAI/blob/95debeafeedaf04f15ea41fe4c9a20a7b06ff933/apps/challenges/aws_utils.py#L216-L222 and https://github.com/Cloud-CV/EvalAI/blob/95debeafeedaf04f15ea41fe4c9a20a7b06ff933/apps/challenges/aws_utils.py#L66-L75

So according to that, the deploymentController param is never called and your code would work if you deleted it entirely. (I will make a pull for that). So indirectly, the deploymentController param was removed entirely. I think it makes more sense now.
@nsjcorps Am I right?
@takitsuse Did you get it as well?

**Comment by nsjcorps** [comment]
@takitsuse  @nikochiko . Where the deployment controller is now. It still gets called. I don't know if  it would work without it. What I do know is that the reason why it is there is so it gets passed to the function that calls it

**Comment by nsjcorps** [comment]
As mentioned before. If you don't use my code and you try it. After resolving the deployment controller issue. You will still run into more issues

**Comment by nsjcorps** [comment]
Or perhaps if we remove the deployment controller totally. It would still work

**Comment by takitsuse** [comment]
@nsjcorps I said "I see" for I see my use of your setup code. Sorry If I didn't make it clear.

**Comment by takitsuse** [comment]
@nsjcorps I'll give it a tentative try. Let's see how it goes.

**Comment by nikochiko** [comment]
@nsjcorps I think that is what you did. VPC_dict is a dict and it is passed as an arg to format. Now, as in the lines I referred above, the deploymentController is not mentioned in the setup like `{deploymentController}` so I don't think it is passed to the function. And therefore, it is equivalent to just removing deploymentController. (I will try to verify this)

**Comment by nsjcorps** [comment]
O alright. It's fine though. We are all trying to do the same thing. @takitsuse

**Comment by nsjcorps** [comment]
@nikochiko I will also try to remove it and see if it works

**Comment by nikochiko** [comment]
Okay then lol, I think I don't have to create a fork on my side. *hush*

**Comment by takitsuse** [comment]
@nikochiko @nsjcorps The test passed.

**Comment by nikochiko** [comment]
Yeah, so that is how it is working. But we have to make sure the deploymentController param isn't required. But I'm not too good with AWS without the documentation, so I am not sure.

**Comment by nsjcorps** [comment]
It's negligible then @takitsuse

**Comment by nsjcorps** [comment]
I will check the docs. But it would have thrown an error if it was right?

**Comment by nsjcorps** [comment]
So I guess I have done all the hardwork for you guys @takitsuse @nikochiko lol ..

**Comment by takitsuse** [comment]
@nsjcorps Yeah I think it is ignorable too.

**Comment by nikochiko** [comment]
@nsjcorps The error being thrown was unexpected in the first place. I think even @KhalidRmb had the same issue. And yes lol, you started the hard work. But just the Travis tests are passing doesn't mean it will behave as expected. I don't think we are testing it in a real AWS environment, so the results may be different then.

**Comment by takitsuse** [comment]
@nsjcorps Thanks for that indeed.
@nikochiko I think the confirmation is too hard for us since we are unfamiliar with AWS stuff.

**Comment by nikochiko** [comment]
@takitsuse Agreed! At least we dug into how @nsjcorps 's fix worked. I will try if it works with @KhalidRmb 's tests as well. Then a lot of work can be saved.

**Comment by takitsuse** [comment]
Pinning @abinav62 because he is in a similar situation in which we were.

**Comment by takitsuse** [comment]
@yashdusing @pushkalkatara @Ram81 @RishabhJain2018 @KhalidRmb
We had a fruitful conversation and now it is resolved.
Please take another look at the changes of this PR.

**Comment by nsjcorps** [comment]
@takitsuse ⭐

**Comment by takitsuse** [comment]
@KhalidRmb Can I have your review please?

**Comment by KhalidRmb** [comment]
Hey @takisuke. Thanks for the PR. This does seem to be a weird issue. I'll investigate it a bit more and get back to you for further review. You can start other tasks if you want to meanwhile. 

This is because, all the commands should be running fine without the modification.

**Comment by takitsuse** [comment]
@KhalidRmb I see. I tentatively abandon this task then. I hope we can find a solution to this issue.

**Comment by takitsuse** [comment]
@KhalidRmb By the way, does abandoning tasks affect the chance of having a good result at GCI? I have abandoned several tasks including this, so I am kind of curious.

**Comment by yashdusing** [comment]
you don't need to remove deploymentController. The problem you've been experiencing is either due to the use of '{{' or improper unpacking while **VPC (newline characters and spaces being in included as key). You could simply call do
`definition['deploymentController'] = { 
 "type": "ECS" 
}`
after eval has been called.

**Comment by yashdusing** [comment]
@takitsuse You can now make the changes as in #2604 for deploymentController

**Comment by takitsuse** [comment]
@yashdusing Thanks for letting me know that, now my build passes without modifying `aws_utils.py` :clap: 
@KhalidRmb Could you take another look at this?

**Comment by takitsuse** [comment]
@vkartik97 @KhalidRmb I rewrote test codes in my original style preserved in https://github.com/Cloud-CV/EvalAI/pull/2524. Could you take another look? Sorry for troubling you.

**Comment by yashdusing** [comment]
@takitsuse could you reorder imports lexicographically ?

File: tests/unit/challenges/test_aws_utils.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        super(BaseAdminCallClass, self).setUp()
+
+        self.challenge3 = Challenge.objects.create(
+            title="Test Challenge 3",
+            creator=self.challenge_host_team,
+            queue="test_queue_3",
+            start_date=timezone.now() - timedelta(days=2),
+            end_date=timezone.now() + timedelta(days=1),
+        )
+        self.ecs_client.create_cluster(clusterName="cluster")
```
Comment (yashdusing) [inline tests/unit/challenges/test_aws_utils.py]: @takitsuse could you update this to `aws_utils.COMMON_SETTINGS_DICT["CLUSTER"]` ?

Comment (takitsuse) [reply tests/unit/challenges/test_aws_utils.py]: I think I already have that code in line 111?

Comment (yashdusing) [reply tests/unit/challenges/test_aws_utils.py]: Okay, nevermind. Could you update the imports ?

Comment (takitsuse) [reply tests/unit/challenges/test_aws_utils.py]: Give me 1 minute, I am updating them.

Comment (takitsuse) [reply tests/unit/challenges/test_aws_utils.py]: @yashdusing Done :+1:

**Comment by yashdusing** [comment]
LGTM :+1:

**Comment by takitsuse** [comment]
@vkartik97 @KhalidRmb Could you also take a look at this?

**Comment by takitsuse** [comment]
@yashdusing Thanks again for the approval!

**Pull request closed without merging**